Issue 181 best transit modes 2

akashkumar2296 opened the pull request:
- added traffic option
- traffic option only called when mode is driving, but it can be selected with any transit (nothing happens)
- retrieve_exact_commutes takes in traffic_option
- get_durations_and_distances takes in traffic_option
- unit tests for both functions changed

Commit:
Testing Done - PR comments complete

Commit:
400 error for traffic_model ??

Commit:
distance wrapper

Commit:
traffic modes, fixed merge conflicts

Commit:
Traffic model complete

Commit:
Ready for PR, function calls/implementations changed

tiusty commented on cocoon/commutes/distance_matrix/distance_wrapper.py:
```diff
@@ -121,13 +121,33 @@ def get_durations_and_distances(self, origins, destinations, mode="driving"):
                                                             units=self.units,
-                                                            mode=mode)
+                                                            mode=mode,
+                                                            departure_time=1543941000,
```
wait i thought we said that we were just going to use best guess and then just change the departure time? Otherwise it isn't accurate. Like if they choose use best guess at 4:30pm it will still include the traffic time, just the best guess with traffic vs pessimestic at 4:30pm which will be like the worst possible traffic. If they choose no traffic is should get the commute time at like 4:00am

tiusty replied:
I guess I see what you did with the traffic switch case but I feel like there should just be one call to the google api and then you just adjust the parameters appropriate. That will make the implementation cleaner I feel.

tiusty commented on cocoon/commutes/distance_matrix/commute_retriever.py:
```diff
@@ -8,7 +8,8 @@
 
 
-def retrieve_exact_commute(origins, destinations, mode):
+def retrieve_exact_commute(origins, destinations, mode=CommuteType.DRIVING, traffic_option=False):
```
not sure if I like the traffic_option variable name cus it is kinda confusing given that it is a boolean. Doing with_traffic as the name I feel is more appropriate

tiusty commented:
Just looking at the bigger picture. How well will this integrate with having transit options? I feel like maybe the retrieve exact commutes should take in a class with all the possible options and then in the get_durations and distance. Options are pulled out that are corresponding to the commute type desired. Otherwise we are going to have a billion different arguments. Though you could delay this to the next PR it might be worth while to set the gorund work so the next PR is easier and cleaner

tiusty commented:
Also you should def add unit tests for the case when traffic is selected. You can use the past tests as a baseline for your new ones. Though def finish making the changes before doing that. I think you should get rid of the if statement for calling the distance matrix and just parse the args

Commit:
PR changes
- Changed name of variable to with_traffic
- Added unit tests with with_traffic call
- Changed if/else to single call with 3:30am for non-traffic option with "best_guess" for both

tiusty commented on cocoon/commutes/distance_matrix/distance_wrapper.py:
```diff
@@ -122,32 +122,28 @@ def get_durations_and_distances(self, origins, destinations, mode="driving", tra
 
         # traffic option set to best_guess (default)
+        # departure time = Tuesday, Dec 4, 2018 3:30am - No Traffic
+        departure_time = 1543894200
```
Any hardcoded values should be in the constants.py file so can you add the constant value to the constants.py with the comment of what that int represents. then just import the value here. 

You can create variables like:

COMMUTE_TIME_WITH_TRAFFIC
COMMUTE_TIME_WITHOUT_TRAFFIC

or something like that

tiusty commented on cocoon/commutes/distance_matrix/distance_wrapper.py:
```diff
@@ -122,32 +122,28 @@ def get_durations_and_distances(self, origins, destinations, mode="driving", tra
+        # departure time = Tuesday, Dec 4, 2018 3:30am - No Traffic
+        departure_time = 1543894200
+
         traffic_model_in="best_guess"
```
This should be in the constants file. Then you actually don't need to create a temp variable since we don't change it. Instead just put the variable striaght into the distance_matrix function.

I.e
TRAFFIC_MODEL_DEFAULT

or something like that

tiusty reviewed (changes requested):
Almost there! It looks good after these changes.

Remember to put the constants variable in the right section and comment what the constant value represents. It should also be in the constants file for that app so in this case it would be commutes/constants.py

Commit:
PR changes
- All values are constants
- Dates are pushed to 2019 (API only accepts departure_time from the future)

tiusty commented on cocoon/commutes/constants.py:
```diff
@@ -20,3 +20,10 @@ class CommuteAccuracy(Enum):
+# API does not accept times from the past
+# With traffic = Tuesday Dec 4, 2019 4:30pm
+# Without traffic = Tuesday Dec 4, 2019 3:30am
+COMMUTE_TIME_WITH_TRAFFIC = 1575477036
```
lol if this is true this is a bug that needs to be addressed cus this will be forgotten about. Can you just do something like dynamically create the day so maybe just use the timezone function where it takes the current day and grabs the next tuesday and then takes that days time at 4:30pm and 3:30am. Cus we cannot rely on people changing it cus when that time hits next year things will break

tiusty reviewed (changes requested):
Critical bug that needs addressing

Commit:
Merge branch 'develop' of https://github.com/tiusty/Cocoon into issue-181-best-transit-modes-2

Commit:
Solution to UNIX time problem

tiusty commented on cocoon/commutes/constants.py:
```diff
@@ -26,4 +27,11 @@ class CommuteAccuracy(Enum):
+TRAFFIC_MODEL_DEFAULT = "best_guess"
+
+# Check if your commute time is in the future, if it is then change both times by a increasing it
+# a year, this makes sure that the time is always in the future
```
this is kinda hacky... could we just use a datetime object to get the exact time we want instead of trying to switch out time?

This will be  a cleaner solution: https://stackoverflow.com/questions/7274267/print-all-day-dates-between-two-dates/7274316

so basically get the current year and then construct the year with date(year, 12, 4). Then you can convert that to unix time. That way we don't even use unix time in the file so it is much easier to read. It is better to not use unix time for the human interface cus it is hard to read.

tiusty reviewed (changes requested):
Last change. I don't like the use of unix time cus it isn't clean

Commit:
Made UNIX timestamps more human readable using datetime

Commit:
All PR changes done

tiusty committed:
Merge branches 'issue-134-refactor-survey' and 'issue-181-best-transit-modes-2' of https://github.com/tiusty/Cocoon into issue-181-best-transit-modes-2

tiusty committed:
[issue-134] Fix function for depature time

tiusty committed:
[issue-134] Fixed traffic time

tiusty committed:
[issue-134] Updated new survey to use traffic option

tiusty committed:
[issue-134] Updated default price weight and commute weight

tiusty committed:
[issue-134] Fixed the price weight question

tiusty committed:
[issue-181] Fix unittests

tiusty committed:
[issue-181] Added unittests

tiusty committed:
[issue-181] Update the traffic time to 5:30pm

tiusty committed:
[issue-181] Update the traffic time to 7:30am

tiusty committed:
[issue-181] Updated time

tiusty committed:
[issue-181] Now the datetime is passed in and the distance matrix does the conversion. This works much better!

tiusty committed:
[issue-181] Uses local timezone

tiusty committed:
[issue-181] Now sets the timezone of the datetime to eastern timezone

tiusty committed:
[issue-181] Now pessimistic is used for computing traffic

tiusty committed:
[issue-181] Change traffic model depending on the type of commute

tiusty committed:
[issue-181] Added comments and unittests

tiusty committed:
[issue-181] Added check in case the duration_in_traffic doesn't exist

tiusty committed:
[issue-181] Switched commute for the afternoon since the destination is the origin

tiusty committed:
[issue-150] Made PR changes
